File: api/schemas/sale.py
from datetime import datetime
from decimal import Decimal

from api.schemas.base import BaseModel
from api.schemas.product import ProductSaleReadDTO


class SaleReadDTO(BaseModel):
    id: int
    # product: ProductSaleReadDTO
    product_id: int
    buy_price: Decimal
    sell_price: Decimal
    quantity: int
    created_by: str
    org_id: str
    created_at: datetime


class SaleCreateDTO(BaseModel):
    product_id: int
    buy_price: Decimal
    sell_price: Decimal
    quantity: int

File: web/src/api/sale.ts
import { api } from './base'

export type SaleCreatePayload = {
  product_id: number
  buy_price: number
  sell_price: number
  quantity: number
}

export type SaleRead = {
  id: number
  product_id: number
  buy_price: string
  sell_price: string
  quantity: number
  created_by: string
  org_id: string
  created_at: string
}

export const createSale = async (payload: SaleCreatePayload, token?: string | null) => {
  const response = await api.post('/sales/', payload, {
    headers: token ? { Authorization: `Bearer ${token}` } : undefined,
  })

  return response.data
}

export const getSales = async (token?: string | null): Promise<SaleRead[]> => {
  const response = await api.get<SaleRead[]>('/sales/', {
    headers: token ? { Authorization: `Bearer ${token}` } : undefined,
  })
  return response.data
}
